### PR TITLE
Fixed #27473 -- Enabled Extract for DurationField.

### PR DESCRIPTION
**Pull request by blueyed**
https://code.djangoproject.com/ticket/27473

### CODE REVIEW

**Comment by jarshwah** [comment]
For throwing an error with the combination of Extract and DurationField, check out:

`django.db.backends.sqlite3.operations.DatabaseOperations.check_expression_support`

**Comment by blueyed** [comment]
@jarshwah 
Thanks.  So I assume this should be done then?  (added it there for now)

Would this also affect e.g. MySQL?  ('has_native_duration_field' seems to be provided only with Oracle and PostgreSQL.)  But I don't know if that is strictly related to / restricted by this feature.

**Comment by blueyed** [comment]
Apparently the fixup causes failures on "spatialite" and "sqlite3" now, which is probably expected.
Please advise how to go on from here.

File: django/db/backends/sqlite3/operations.py [hunk 1]
```diff
@@ -27,7 +27,7 @@ def bulk_batch_size(self, fields, objs):
         return (limit // len(fields)) if len(fields) > 0 else len(objs)
 
     def check_expression_support(self, expression):
-        bad_fields = (fields.DateField, fields.DateTimeField, fields.TimeField)
+        bad_fields = (fields.DateField, fields.DateTimeField, fields.TimeField, fields.DurationField)
```
Comment (jarshwah) [inline django/db/backends/sqlite3/operations.py]: Sorry, I didn't provide enough context here for how I imagined this to be fixed. The existing logic here enforces that aggregations can't be used with these fields. Aggregating a duration field is fine on sqlite.

Instead, you should write a check here like:
```python
try:
    if isinstance(expression, Extract) and isinstance(expression.output_field, fields.DurationField):
        raise NotImplementedError("Extract .. on DurationField fails on sqlite")
except FieldError:
    pass
```

Does that make sense?

Comment (blueyed) [inline django/db/backends/sqlite3/operations.py]: @jarshwah 
Thanks for your review, I've added your suggested change.

**Comment by blueyed** [comment]
@timgraham 
Thanks for your review!

> How about a helpful error message for databases like MySQL that have `has_native_duration_field=False`

Makes sense.
Should it be a `ValueError` in `Extract.as_sql`?  (I've done that for now)

**Comment by blueyed** [comment]
> Should it be a ValueError in Extract.as_sql? (I've done that for now)

That would make the SQLite specific check (in `check_expression_support`) obsolete - at least for the tests I have currently.

**Comment by timgraham** [comment]
By the way, it's better to make all the adjustments locally and then push once to GitHub since each push creates a build on the CI server. For a small commit like this, it's fine to amend the commit rather than creating a bunch of fixup commits.

**Comment by blueyed** [comment]
@timgraham 
Fair point.
But I'd rather keep the history here for now, since I'm not sure if "Remove SQLite specific check from check_expression_support" is the way to go, given that it was added first.  /cc @jarshwah

**Comment by blueyed** [comment]
Rebased.

**Comment by blueyed** [comment]
The "docs" build failure seems unrelated (https://djangoci.com/job/docs/16778/console).

Please review it again, and then I will squash it.

**Comment by timgraham** [comment]
Seems like some documentation updates are required (ref/models/database-functions.html and in the 2.0 release notes). The new test is failing on Oracle with ORA-00907: missing right parenthesis. You can try https://code.djangoproject.com/wiki/OracleTestSetup or maybe @felixxm will help.

Uncheck "Patch needs improvement" on the ticket when it's ready for review.

File: tests/db_functions/test_datetime.py [hunk 2]
```diff
@@ -198,6 +200,37 @@ def test_extract_func(self):
...
+        start_datetime = microsecond_support(datetime(2015, 6, 15, 14, 30, 50, 321))
+        end_datetime = microsecond_support(datetime(2016, 6, 15, 14, 10, 50, 123))
+        if settings.USE_TZ:
+            start_datetime = timezone.make_aware(start_datetime, is_dst=False)
+            end_datetime = timezone.make_aware(end_datetime, is_dst=False)
+        self.create_model(start_datetime, end_datetime)
+        self.create_model(end_datetime, start_datetime)
+
+        self.assertQuerysetEqual(
+            DTModel.objects.annotate(extracted=Extract('duration', 'epoch')).order_by('start_datetime'),
```
Comment (felixxm) [inline tests/db_functions/test_datetime.py]: You cannot use a `EXTRACT` with `EPOCH` on Oracle (see [doc](https://docs.oracle.com/cd/B28359_01/server.111/b28286/functions052.htm#SQLRF00639)). Please use another lookup.

Comment (blueyed) [inline tests/db_functions/test_datetime.py]: @felixxm 
Should this test get changed then?
Can you suggest something?
(Sorry, this PR is quite old, I need to look into it in total again first)

File: tests/db_functions/test_datetime.py [hunk 3]
```diff
@@ -198,6 +200,37 @@ def test_extract_func(self):
...
+            lambda m: (m.start_datetime, m.extracted)
+        )
+        self.assertEqual(DTModel.objects.annotate(duration_days=Extract('duration', 'day')).filter(
+            duration_days__gt=200).count(), 1)
+
+    @skipIfDBFeature('has_native_duration_field')
+    def test_extract_duration_without_native_duration_field(self):
+        msg = "Extract() requires native DurationField database support."
+        with self.assertRaisesMessage(ValueError, msg):
+            list(DTModel.objects.annotate(extracted=Extract('duration', 'epoch')))
```
Comment (felixxm) [inline tests/db_functions/test_datetime.py]: Use another lookup instead of `epoch` e.g. `second`.

File: tests/db_functions/test_datetime.py (hunk 2 above)
Comment (felixxm) [inline tests/db_functions/test_datetime.py]: I think that this part can be dropped:
```diff
diff --git a/tests/db_functions/test_datetime.py b/tests/db_functions/test_datetime.py
index 51dbcb6..3560a76 100644
--- a/tests/db_functions/test_datetime.py
+++ b/tests/db_functions/test_datetime.py
@@ -211,12 +211,6 @@ class DateFunctionTests(TestCase):
         self.create_model(end_datetime, start_datetime)
 
         self.assertQuerysetEqual(
-            DTModel.objects.annotate(extracted=Extract('duration', 'epoch')).order_by('start_datetime'),
-            [(start_datetime, int((end_datetime - start_datetime).total_seconds())),
-             (end_datetime, int((start_datetime - end_datetime).total_seconds()))],
-            lambda m: (m.start_datetime, m.extracted)
-        )
-        self.assertQuerysetEqual(
```

**Comment by blueyed** [comment]
Incorporated feedback/comments, and added it to docs in `docs/ref/models/database-functions.txt`.

**Comment by timgraham** [comment]
Add a `.. versionchanged:: 2.0` annotation as described in [documenting new features](https://docs.djangoproject.com/en/dev/internals/contributing/writing-documentation/#documenting-new-features). Perhaps the fact that this doesn't work on all databases should also be documented? Also, include a release note in `docs/releases/2.0.txt`.

**Comment by blueyed** [comment]
@timgraham 
Thanks for merging!

>  timgraham deleted a comment from blueyed an hour ago

?

**Pull request merged**